abdiff/core/collate_ab_transforms.py

```diff
@@ -207,6 +212,30 @@ def get_joined_batches_iter(dataset_directory: str) -> Generator[pa.RecordBatch]
                     break
 
 
+def validate_output(dataset_path: str) -> None:
+    """Validate the output of collate_ab_transforms.
+
+    This function checks whether the collated dataset is empty
+    and whether any or both 'record_*' columns are empty.
+    """
+    collated_dataset = ds.dataset(dataset_path)
+    collated_table = collated_dataset.to_table(columns=["record_a", "record_b"])
```

Comment:
I hate to say it, but I think this might need some rework.  Calling `collated_dataset.to_table(...)` will pull the entire dataset into memory.

I think checking if the dataset is entirely empty could be handled by dataset methods like `.take(10)` or `.head(10)`, as you'd expect those to return zero results if so, but it's only getting the first few records which is memory safe.

As for checking if `record_a` or `record_b` is totally empty... it's possible that a DuckDB query could be an efficiente way to do this?

In fact, DuckDB query(s) could be used for _all_ of this, with the expectation that DuckDB is memory efficient and would not load the entire dataset into memory.

Reply:
This wouldn't have been obvious to me either, so thank you very much for pointing it out!

Reply:
@ghukill Just wanted to clarify (and make sure I understand as well): 
The code highlighted above would load only the **selected columns** specified in `.to_table(columns=[...])`, right? From the [PyArrow docs](https://arrow.apache.org/docs/python/dataset.html#dataset-discovery) it reads:

> Creating a [Dataset](https://arrow.apache.org/docs/python/generated/pyarrow.dataset.Dataset.html#pyarrow.dataset.Dataset) object does not begin reading the data itself. If needed, it only crawls the directory to find all the files:

However, what you're saying is that even if we are only loading in a subset of columns `record_a` and `record_b`, that can _still_ consume a lot of memory if we are diff-ing a lot of input files. 

Let me know what ya'll think of [the update to use DuckDB queries](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/35/files#diff-1cf5419c409425ac3b8275acb901b7c68cc1131a5b24d5846f477eb967681e55R221-R262). I think I personally prefer the concise syntax with the previous method, _but_ I definitely want to be mindful of memory use!

Reply:
You're right that loading a dataset does not bring anything into memory until other operations are performed, 100%.  So we're safe until the `to_table()` call.

> However, what you're saying is that even if we are only loading in a subset of columns record_a and record_b, that can still consume a lot of memory if we are diff-ing a lot of input files.

That's also correct!  But, keep in mind that `record_a` and `record_b` are binary blobs of the _full_ records, so they are hold relatively large values per row.  

Assume the collate dataset held the A and B versions of records from `alma`, maybe around 3 million records.  The transformed records from A or B (just one) is about 14gb in JSON files, probably somewhere around 10gb in parquet (compressed).  Regardless of compression, bring in _both_ A and B versions (`record_a` and `record_b` columns), which be potentially around 20gb in memory.  Not only would this be quite slow, it would often overwhelm memory (I believe all of our machines are 16gb).  

To summarize: even though we subset the columns, and it's lazy loaded until `to_table()`, the size of those columns makes it an very memory-expensive operation.

Interestingly, when calcing metrics, we _do_ call something similar to `to_table()` on the entire dataset, but we are filtering down to columns with very, very small values.  So even if 10-20 million rows, it may only be 50-100mb in memory, tops.  The nature of the columns we select are critical to deciding if we can safely bring into memory of not.